### PR TITLE
Fix cookie parsing issues

### LINKED ISSUE #2683: CookieError on invalid cookie names
## Long story short

aiohttp servers fail with 500 Internal error on requests where Cookie header is present and cookie name contains invalid symbols for cookie name. aiohttp uses SimpleCookie from python's http.cookies package. Although SimpleCookie follows standards and supposedly correctly rejects such cookie names, real world is much messier and such cookies are seen frequently.

For the reference, RFC2109 states that cookie attribute (name) is a `token`:

    The two state management headers, Set-Cookie and Cookie, have common
    syntactic properties involving attribute-value pairs.  The following
    grammar uses the notation, and tokens DIGIT (decimal digits) and
    token (informally, a sequence of non-special, non-white space
    characters) from the HTTP/1.1 specification [RFC 2068] to describe
    their syntax.
    
    av-pairs        =       av-pair *(";" av-pair)
    av-pair         =       attr ["=" value]        ; optional value
    attr            =       token

RFC2616 defines `token` as:

    token          = 1*<any CHAR except CTLs or separators>
    separators     = "(" | ")" | "<" | ">" | "@"
                   | "," | ";" | ":" | "\" | <">
                   | "/" | "[" | "]" | "?" | "="
                   | "{" | "}" | SP | HT

So any cookie attribute containing one of `( ) < > @ , ; : \ " / [ ] ? = { } SP HT` is considered invalid by SimpleCookie. Django uses own cookie parsing because of similar reasons.

## Expected behaviour

Such invalid cookie should still be accepted as they are tolerated by browsers and most web servers.

## Actual behaviour

500 Internal server error caused by exception CookieError.

## Steps to reproduce

Given sample http server:

    from aiohttp import web
    
    async def handle(request):
        text = ""
        for k, v in request.cookies:
            text += '\n{}\t{}'.format(k, v)
        return web.Response(text=text)
    
    app = web.Application()
    app.router.add_get('/', handle)
    app.router.add_get('/{name}', handle)
    
    web.run_app(app)

When one executes `curl -H 'Cookie: ISAWPLB{DB45DF86-F806-407C-932C-D52A60E4019E}=x' -v http://localhost:8080/` against this server it receives 500 internal server error. On the server side a traceback is visible:

    Error handling request
    Traceback (most recent call last):
      File "/usr/local/lib/python3.5/dist-packages/aiohttp/helpers.py", line 554, in __get__
        return inst._cache[self.name]
    KeyError: 'cookies'
    
    During handling of the above exception, another exception occurred:
    
    Traceback (most recent call last):
      File "/usr/local/lib/python3.5/dist-packages/aiohttp/web_protocol.py", line 410, in start
        resp = yield from self._request_handler(request)
      File "/usr/local/lib/python3.5/dist-packages/aiohttp/web.py", line 325, in _handle
        resp = yield from handler(request)
      File "/usr/local/lib/python3.5/dist-packages/aiohttp/web_middlewares.py", line 93, in impl
        return (yield from handler(request))
      File "./test.py", line 7, in handle
        for k, v in request.cookies:
      File "/usr/local/lib/python3.5/dist-packages/aiohttp/helpers.py", line 556, in __get__
        val = self.wrapped(inst)
      File "/usr/local/lib/python3.5/dist-packages/aiohttp/web_request.py", line 413, in cookies
        parsed = SimpleCookie(raw)
      File "/usr/lib/python3.5/http/cookies.py", line 507, in __init__
        self.load(input)
      File "/usr/lib/python3.5/http/cookies.py", line 556, in load
        self.__parse_string(rawdata)
      File "/usr/lib/python3.5/http/cookies.py", line 620, in __parse_string
        self.__set(key, rval, cval)
      File "/usr/lib/python3.5/http/cookies.py", line 512, in __set
        M.set(key, real_value, coded_value)
      File "/usr/lib/python3.5/http/cookies.py", line 380, in set
        raise CookieError('Illegal key %r' % (key,))
    http.cookies.CookieError: Illegal key 'ISAWPLB{DB45DF86-F806-407C-932C-D52A60E4019E}'

## Your environment

Tested on Ubuntu 16.04, aiohttp version 2.3.9 installed through pip.

### LINKED ISSUE #4327: Compliance with rfc6265 for cookie expires dates
## Long story short

Compliance with rfc6265 for cookie expires functionality.  Took a while yesterday trying to work out why some client code wasn't working correctly, eventually tracked it down to cookie handling in aiohttp; well in pythons http.cookies.  The server was returning an old but should be valid date format.

## Expected behaviour

The specific problems I had are related to expires.  Digging through the rfc's my current understanding is that expires as defined in rfc6265 states that the date format should be a sane-cookie-date as described in https://tools.ietf.org/html/rfc2616#section-3.3.1
```
  Sun, 06 Nov 1994 08:49:37 GMT  ; RFC 822, updated by RFC 1123
  Sunday, 06-Nov-94 08:49:37 GMT ; RFC 850, obsoleted by RFC 1036
  Sun Nov  6 08:49:37 1994       ; ANSI C's asctime() format
```

Problem is that aiohttp doesn't handle all these options and it appears to strip cookies that have two of these formats.

## Steps to reproduce

```python
import sys
import http.server
import socketserver

class MainHandler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        self.send_response(200)
        self.send_header('Set-Cookie', 'verify1=verify1; Path=/;')
        self.send_header('Set-Cookie', 'rfc822=rfc822value; expires=Fri, 06 Nov 2020 08:49:37 GMT; Max-Age=31449600; Path=/;')
        self.send_header('Set-Cookie', 'rfc850=rfc850value; expires=Friday, 06-Nov-20 08:49:37 GMT; Max-Age=31449600; Path=/;')
        self.send_header('Set-Cookie', 'ansic=ansicvalue; expires=Fri Nov  6 08:49:37 2020; Max-Age=31449600; Path=/;')
        self.send_header('Set-Cookie', 'common=commonvalue; expires=Fri, 06-Nov-2020 16:01:51 GMT; Max-Age=31449600; Path=/;')
        self.send_header('Set-Cookie', 'expired=expiredvalue; expires=Fri, 06 Nov 1999 14:18:35 GMT; Path=/;')
        self.end_headers()
        self.wfile.write(b'cookies!')

def main(port):
    with socketserver.TCPServer(("", port), MainHandler) as httpd:
        print("serving at port", port)
        httpd.serve_forever()

if __name__ == '__main__':
    main(int(sys.argv[1]))
```

This is a simple python server to return the different values, I included the common type with 4 digit years as everywhere seems to use it.

If I run a request with aiohttp

```python
import aiohttp
import asyncio

async def main():
    async with aiohttp.ClientSession() as session:
        async with session.get('http://localhost:8000') as response:
            await response.text()

            for c in response.cookies:
                print(c)
            print('in session')
            for c in session.cookie_jar:
                print(c)

if __name__ == '__main__':
    loop = asyncio.get_event_loop()
    loop.run_until_complete(main())
```

## Actual behaviour

I will get output along the lines of

```
verify1
rfc822
common
expired
in session
Set-Cookie: verify1=verify1; Domain=localhost; Path=/
Set-Cookie: rfc822=rfc822value; Domain=localhost; expires=Fri, 06 Nov 2020 08:49:37 GMT; Max-Age=31449600; Path=/
Set-Cookie: common=commonvalue; Domain=localhost; expires=Fri, 06-Nov-2020 16:01:51 GMT; Max-Age=31449600; Path=/
```

rfc850 and ansic dates are excluded during response parsing.

I'm not sure if supporting these formats is important to you, I'm going to get the server we are connecting to to switch formats aliviating the problem for me specifically.  Just thought I would report it anyway.

OS: MacOSX
Python: 3.7
aiohttp: 3.6.2

### LINKED ISSUE #4493: Allow RFC 2822-compliant dates for the `expires` cookie directive
## Long story short

I was using a `ClientSession` with an API that replied with `Set-Cookie`, but the date format use for `expires` was not [RFC 2616](https://tools.ietf.org/html/rfc2616#section-3.3.1)-compliant, but [RFC 2822](https://tools.ietf.org/html/rfc2822#section-3.3)-compliant: the timezone was not `GMT`, but `-0000`.
More concretely, the following cookie fails:

    hello=world; expires=Wed, 15 Jan 2020 09:45:07 -0000

While this one works:

    hello=world; expires=Wed, 15 Jan 2020 09:45:07 GMT

## Expected behaviour

I expect that a cookie set via `Set-Cookie` and with an `expires` field compliant with RFC 2822 is correctly set in the session's cookies.

## Actual behaviour

The cookie is passed to `http.cookies.SimpleCookie`, and at one point goes through the `_CookiePattern` regex that explicitly requires date fields to end with `GMT`.

`http.cookies:434`:

```python
_CookiePattern = re.compile(r"""
    \s*                            # Optional whitespace at start of cookie
    (?P<key>                       # Start of group 'key'
    [""" + _LegalKeyChars + r"""]+?   # Any word of at least one letter
    )                              # End of group 'key'
    (                              # Optional group: there may not be a value.
    \s*=\s*                          # Equal Sign
    (?P<val>                         # Start of group 'val'
    "(?:[^\\"]|\\.)*"                  # Any doublequoted string
    |                                  # or
    \w{3},\s[\w\d\s-]{9,11}\s[\d:]{8}\sGMT  # Special case for "expires" attr
    |                                  # or
    [""" + _LegalValueChars + r"""]*      # Any word or empty string
    )                                # End of group 'val'
    )?                             # End of optional value group
    \s*                            # Any number of spaces.
    (\s+|;|$)                      # Ending either at space, semicolon, or EOS.
    """, re.ASCII | re.VERBOSE)    # re.ASCII may be removed if safe.
```

Line `444`:

```
\w{3},\s[\w\d\s-]{9,11}\s[\d:]{8}\sGMT  # Special case for "expires" attr
```

## Steps to reproduce

Please see [this StackOverflow post](https://stackoverflow.com/questions/59644106/aiohttp-clientsession-does-not-set-cookie-with-expires-key) where I include an MWE showing the behaviour, as well as an MWE with `requests` which shows the expected behaviour. 

## Your environment

```bash
$ python --version                                      
Python 3.7.4
$ python -c "import aiohttp; print(aiohttp.__version__)"
3.6.2

```

### LINKED ISSUE #5397: [Client] Inconsistent quoted cookie handling
🐞 **Describe the bug**

Quoted cookies are unquoted if domain attribute is not set.

💡 **To Reproduce**
Run this code:
```python
import aiohttp
import asyncio
from http.cookies import SimpleCookie


async def request_with_cookie(session_cookie: str, request_cookie: str):
    session_cookie = SimpleCookie(session_cookie)
    request_cookie = SimpleCookie(request_cookie)

    async with aiohttp.ClientSession(cookies=session_cookie) as session:
        async with session.get(
            'https://httpbin.org/headers',
            cookies=request_cookie,
        ) as response:
            html = await response.text()
            print("Body:", html)


async def main():
    await request_with_cookie(
        'sess="quoted_value"; domain=.httpbin.org',
        'req="quoted_value"'
    )

```

Outputs
```
Body: {
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate", 
    "Cookie": "req=quoted_value; sess=\"quoted_value\"", 
    "Host": "httpbin.org", 
    "User-Agent": "Python/3.8 aiohttp/3.7.3", 
    "X-Amzn-Trace-Id": "Root=1-5ffc692a-7ee8f7946bd59b5f3dc67407"
  }
}
```
Cookie with `domain` attribute is quoted, and cookie without `domain` attribute is not quoted

💡 **Expected behavior**

I think if user inputs a quoted cookie, the quotes should not be dropped.
At least it should be consistent, and not dependent on `domain` attribute.

📋 **Logs/tracebacks**

```python-traceback (paste your traceback in the next line)

```

📋 **Your version of the Python**

Python 3.8.5

📋 **Your version of the aiohttp/yarl/multidict distributions**

```console
$ python -m pip show aiohttp
Version: 3.7.3
...
```
```console
$ python -m pip show multidict
Version: 4.7.6
...
```
```console
$ python -m pip show yarl
Version: 1.4.2

...
```

📋 **Additional context**

Additional code to what I think is the cause
```python
from http.cookies import SimpleCookie
from aiohttp import CookieJar
import yarl


def process_cookie_string(cookie_str: str) -> None:
    cookies = SimpleCookie(cookie_str)
    # Done internally by aiohttp - filtering cookies by domain.
    # https://github.com/aio-libs/aiohttp/blob/v3.7.3/aiohttp/client.py#L477
    tmp_cookie_jar = CookieJar()
    tmp_cookie_jar.update_cookies(cookies)
    req_cookies = tmp_cookie_jar.filter_cookies(yarl.URL('https://www.example.com/'))

    for name, value in req_cookies.items():
        print(f"{name=}, {value.value=}, {value.coded_value=}")

if __name__ == '__main__':
    process_cookie_string('name="value"')
    process_cookie_string('name="value"; Domain=example.com')

```

```
name='name', value.value='value', value.coded_value='value'
name='name', value.value='value', value.coded_value='"value"'
```

Also, these example might be a bit of a stretch because official docs say that cookies should be passed as `dict`.
But is there any reason why I shouldn't be able to send quoted cookies?

### LINKED ISSUE #7993: aiohttp.web fails to parse cookies after a cookie with quotes
### Describe the bug

Cookie values after a double quote `"` are not parsed. All subsequent cookies in the request are silently dropped.

### To Reproduce

1. setup the following simple aiohttp server:
```python
#!/usr/bin/env python
from aiohttp import web
import json

async def hello(request):
    return web.Response(text=str(request.cookies))

app = web.Application()
app.add_routes([web.get("/", hello)])

web.run_app(app)
```
2. run the server
3. make a request like this which has cookies with a double quote in the value: `curl http://localhost:8080 -H 'Cookie: baz="qux; foo=bar;'`
4. notice that the response body/cookies are empty not `{'baz'='"qux', 'foo': 'bar'}`

### Expected behavior

cookie value would be parsed and returned with a double quote in the value, subsequent cookies would also not be silently dropped

### Logs/tracebacks
correct behavior
```sh
> curl http://localhost:8080 -H 'Cookie: foo=bar;'
{'foo': 'bar'}
> curl http://localhost:8080 -H 'Cookie: foo=bar; baz=qux;'
{'foo': 'bar', 'baz': 'qux'}
> curl http://localhost:8080 -H 'Cookie: foo=bar; baz=qux; foo2=bar2'
{'foo': 'bar', 'baz': 'qux', 'foo2': 'bar2'}
```

the bug:
```sh
> curl http://localhost:8080 -H 'Cookie: foo=bar; baz="qux; foo2=bar2'
{'foo': 'bar'}
```

### Python Version

```console
$ python --version
Python 3.11.5
```

### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.9.1
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author:
Author-email:
License: Apache 2
Location: /Users/nick/.asdf/installs/python/3.11.5/lib/python3.11/site-packages
Requires: aiosignal, attrs, frozenlist, multidict, yarl
Required-by: openai
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.4
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /Users/nick/.asdf/installs/python/3.11.5/lib/python3.11/site-packages
Requires:
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.9.4
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache-2.0
Location: /Users/nick/.asdf/installs/python/3.11.5/lib/python3.11/site-packages
Requires: idna, multidict
Required-by: aiohttp
```

### OS

macos 14.1.1 (23B81)

### Related component

Server

### Additional context

We recently deployed a service using aiohttp and noticed that some users would be "logged in" i.e. have a valid session by our main server (webapp2), and not be logged in on our async aiohttp server. It was really hard to narrow down the specfic issue but eventually we found that a requests with a cookie value containing json before our session cookie would cause the user to appear logged out to our aiohttp server

other servers handle the data more robustly for example flask, will correctly parse the 2nd cookie and cookies after it:
```python
#!/usr/bin/env python3

from flask import Flask, request

app = Flask(__name__)

@app.route("/")
def hello():
    return str(request.cookies)

def main():
    app.run(debug=True)

if __name__ == '__main__':
    main()
```

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### PR DESCRIPTION
**Pull request by bdraco**
## Summary

This PR fixes multiple long-standing cookie parsing issues by implementing a more lenient cookie parser that handles real-world cookies while maintaining compatibility with the HTTP cookie specification.

### Issues Fixed
- Fixes #2683
- Fixes #7993
- Fixes #5397
- Fixes #4327
- Fixes #4493

### 1. Use `parse_cookie_headers` for parsing cookie headers
Replaced Python's strict `SimpleCookie` with the more lenient `parse_cookie_headers` function in multiple places:

**Server-side (aiohttp/web_request.py):**
- Modified cookie parsing in incoming requests to accept cookies with special characters (fixes #2683)
- Prevents 500 errors when encountering cookies with characters like `{}[]()` that are commonly used by real-world websites

**Client-side:**
- **aiohttp/client_reqrep.py**: Modified cookie parsing in two places:
  - **Response cookies**: Uses `parse_cookie_headers` to handle malformed cookies gracefully
    - Maintains backward compatibility by still exposing `response.cookies` as a `SimpleCookie` object
    - Uses `parse_cookie_headers` for the actual parsing, then updates the SimpleCookie with the results
  - **Request cookies**: Updated `update_cookies()` method to use `parse_cookie_headers` when parsing existing Cookie headers
    - Prevents errors when updating cookies if the existing Cookie header contains special characters
- **aiohttp/abc.py**: Simplified `update_cookies_from_headers()` in AbstractCookieJar
  - Removed try/except blocks since `parse_cookie_headers` handles errors internally without raising exceptions
  - Used by all cookie jar implementations (CookieJar, DummyCookieJar) for consistent cookie handling

### 2. Updated logger usage
- Changed `client_logger` to `internal_logger` in `parse_cookie_headers` since the function is now used by both client and server code

### 3. Fixed inconsistent cookie quoting behavior in CookieJar
- Shared cookies (domain="", path="") were not respecting the `quote_cookie` setting
- Non-shared cookies were properly respecting `quote_cookie`, but shared cookies were always sent unquoted
- Fixed by using the same `_build_morsel()` method for both shared and non-shared cookies
- This ensures all cookies respect the `self._quote_cookie` setting consistently

### 4. Refactored cookie morsel building in CookieJar
- Extracted cookie morsel creation into a dedicated `_build_morsel()` method
- Uses the new helper functions from `_cookie_helpers` module:
  - `make_non_quoted_morsel()` when `quote_cookie=False`
  - `preserve_morsel_with_coded_value()` when cookie is already properly quoted
  - `make_quoted_morsel()` when cookie needs quoting
- Simplified the code by always using `BaseCookie` for the filtered result to ensure our quoting is always respected.

### 5. Comprehensive test coverage
Added extensive tests to ensure:
- Cookies with special characters are accepted
- Real-world cookie formats work correctly
- Quoted cookie values are handled consistently
- Edge cases are handled properly

## Technical Details

- More tolerant parsing of cookie names
- Consistent unquoting of cookie values
- Maintains backward compatibility
- Gracefully handles malformed cookies
- Ensures consistent behavior for all cookies regardless of domain/path

The implementation aims to make aiohttp more robust when handling cookies from various web services and frameworks.

## Are there changes in behavior for the user?

More lenient parser.

## Is it a substantial burden for the maintainers to support this?

We have to maintain a cookie parser, however its not as bad as I was expecting since our use case is more narrow than everything `SimpleCookie` does

### CODE REVIEW

File: aiohttp/helpers.py [hunk 1]
```diff
@@ -77,6 +79,66 @@
...
 COOKIE_MAX_LENGTH = 4096
 
+# Cookie parsing constants
+# Allow more characters in cookie names to handle real-world cookies
+# that don't strictly follow RFC standards (fixes #2683)
+# RFC 6265 defines cookie-name token as per RFC 2616 Section 2.2,
+# but many servers send cookies with characters like {} [] () etc.
+# This makes the cookie parser more tolerant of real-world cookies
+# while still providing some validation to catch obviously malformed names.
+_COOKIE_NAME_RE = re.compile(r"^[!#$%&\'()*+\-./0-9:<=>?@A-Z\[\]^_`a-z{|}~]+$")
```
Comment (Dreamsorcerer) [inline aiohttp/helpers.py]: If we're following RFC 6265, then we shouldn't have these regexes at all:
https://datatracker.ietf.org/doc/html/rfc6265#section-5.2

Parser should be something more like:
```
for name_value_pair in set_cookie_header.split(";"):
    name, sep, value = name_value_pair.partition("=")
    if not sep:
        continue
    name = name.strip()
    if not name:
        continue
    value = value.strip()
```

Comment (bdraco) [reply aiohttp/helpers.py]: Good call. I was trying to be bug for bug compat with SimpleCookie but more forgiving. Let's make it simpler

Comment (bdraco) [reply aiohttp/helpers.py]: = won't be there for Secure or Httponly though...

Comment (bdraco) [reply aiohttp/helpers.py]: We also have a problem with name="coo;ie" than

Comment (bdraco) [reply aiohttp/helpers.py]: That means we need to write two versions one for parsing Set-Cookie and one for Cookie

We also can't follow the RFC exactly either.

I may need to start over

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: I'm also wondering if we should be parsing multiple headers:
https://github.com/aio-libs/aiohttp/blob/21d640dc4b046be18bcb42aad8fa5445198fd219/aiohttp/client_reqrep.py#L494

For creation, it's very clear that the agent MUST not create multiple headers. But, it doesn't explicitly describe any behaviour that should be followed if we do receive multiple headers.

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: Oh wait, maybe that was Cookie and not Set-Cookie, think I confused myself.

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: > = won't be there for Secure or Httponly though...

Yeah, I didn't go through the second half of the algorithm, so that's not right. There shouldn't be a loop.

Comment (bdraco) [reply aiohttp/helpers.py]: We have tons of examples of multiple headers.  

We also need two parsers, one for Set-Cookie: and one for Cookie:

We can't use the simple cookie approach at all if we want to be compliant with https://datatracker.ietf.org/doc/html/rfc6265#section-5.2

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: Something more like:
```
name_value_pair, *unparsed_attributes = set_cookie_header.split(";")
name, sep, value = name_value_pair.partition("=")
if not sep:
    return
name = name.strip()
if not name:
    return
value = value.strip()

for cookie_av in unparsed_atttributes:
    attribute_name, _, attribute_value = cookie_av.partition("=")
    attribute_name = attribute_name.strip()
    attribute_value = attribute_value.strip()

    #   6.  Process the attribute-name and attribute-value according to the
    #  requirements in the following subsections.  (Notice that
    #  attributes with unrecognized attribute-names are ignored.)
```

Comment (bdraco) [reply aiohttp/helpers.py]: When I implement https://datatracker.ietf.org/doc/html/rfc6265#section-5.2 a significant amount of cases fail

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: > That means we need to write two versions one for parsing Set-Cookie and one for Cookie

Yes, so actually, I think we may need to keep the regexes for Cookie parsing. Cookie should be more well defined, as they mostly come from browsers with a reasonable consensus on formatting. Set-Cookie comes from all sorts of servers and applications that fail to follow a consistent set of rules.

Comment (bdraco) [reply aiohttp/helpers.py]: Going to go down the path of seeing if we can make the parser work. last working commit is https://github.com/aio-libs/aiohttp/pull/11112/commits/a1229ef80feaad28f37b3e6130aa7a173ab17689

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: > When I implement https://datatracker.ietf.org/doc/html/rfc6265#section-5.2 a significant amount of cases fail

Probably to be expected as some of those tests are testing that the behaviour conforms to an obsoleted RFC. The tests would need to updated/removed.

Comment (bdraco) [reply aiohttp/helpers.py]: I don't see how we can ever make https://datatracker.ietf.org/doc/html/rfc6265#section-5.2 work for us as we specifically have tests for missing semi-colons and all sort of other issues

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: > We can't use the simple cookie approach at all if we want to be compliant with https://datatracker.ietf.org/doc/html/rfc6265#section-5.2

Yes, I mentioned that earlier on another PR. We previously discussed this somewhere and suggested that completely removing http.cookies from our code would be the only to get cookie behaviour working correctly.

It's likely to be a notable breaking change, so wouldn't be backported.

Comment (bdraco) [reply aiohttp/helpers.py]: This works but its absolutely horrible.

```

def parse_set_cookie_header(
    set_cookie_string: str,
) -> Optional[Tuple[str, Morsel[str]]]:
    """
    Parse a single Set-Cookie header according to RFC 6265 Section 5.2.

    This implementation follows the algorithm specified in RFC 6265 for parsing
    set-cookie-string values, ensuring proper handling of cookie names, values,
    and attributes.

    Args:
        set_cookie_string: A single Set-Cookie header value

    Returns:
        A tuple of (name, morsel) if parsing succeeds, None otherwise
    """
    if not set_cookie_string:
        return None

    # Step 1: Split into name-value-pair and unparsed-attributes
    semicolon_pos = set_cookie_string.find(";")
    if semicolon_pos == -1:
        name_value_pair = set_cookie_string
        unparsed_attributes = ""
    else:
        name_value_pair = set_cookie_string[:semicolon_pos]
        unparsed_attributes = set_cookie_string[semicolon_pos:]

    # Step 2: Check for = character in name-value-pair
    equals_pos = name_value_pair.find("=")
    if equals_pos == -1:
        # Ignore the set-cookie-string entirely
        return None

    # Step 3: Split into name and value
    name = name_value_pair[:equals_pos]
    value = name_value_pair[equals_pos + 1 :]

    # Step 4: Remove leading/trailing WSP from name and value
    name = name.strip()
    value = value.strip()

    # Step 5: Check if name is empty
    if not name:
        # Ignore the set-cookie-string entirely
        return None

    # Step 6: cookie-name is name, cookie-value is value
    # Validate the cookie name
    if name.lower() in _COOKIE_KNOWN_ATTRS or not _COOKIE_NAME_RE.match(name):
        internal_logger.warning(
            "Can not load response cookies: Illegal cookie name %r", name
        )
        return None

    # Create the morsel with the parsed name and value
    morsel = Morsel()
    # Unquote the value if it's quoted
    unquoted_value = _unquote(value)
    _set_validated_morsel_values(morsel, name, unquoted_value)

    # Parse unparsed-attributes according to the algorithm
    pos = 0
    while pos < len(unparsed_attributes):
        # Step 1: Skip if empty
        if pos >= len(unparsed_attributes):
            break

        # Step 2: Discard the first character (semicolon)
        if unparsed_attributes[pos] == ";":
            pos += 1
            if pos >= len(unparsed_attributes):
                break

        # Skip any whitespace after semicolon
        while pos < len(unparsed_attributes) and unparsed_attributes[pos] in " \t":
            pos += 1

        if pos >= len(unparsed_attributes):
            break

        # Step 3: Find the next semicolon or end of string
        next_semicolon = unparsed_attributes.find(";", pos)
        if next_semicolon == -1:
            cookie_av = unparsed_attributes[pos:]
            pos = len(unparsed_attributes)
        else:
            cookie_av = unparsed_attributes[pos:next_semicolon]
            pos = next_semicolon

        # Step 4: Check for = in cookie-av
        equals_pos = cookie_av.find("=")
        if equals_pos == -1:
            attribute_name = cookie_av
            attribute_value = ""
        else:
            attribute_name = cookie_av[:equals_pos]
            attribute_value = cookie_av[equals_pos + 1 :]

        # Step 5: Remove leading/trailing WSP
        attribute_name = attribute_name.strip()
        attribute_value = attribute_value.strip()

        # Step 6: Process the attribute
        if attribute_name:
            attr_lower = attribute_name.lower()
            # Handle $ prefixed attributes (legacy cookie format)
            if attr_lower.startswith("$"):
                attr_lower = attr_lower[1:]

            if attr_lower in _COOKIE_KNOWN_ATTRS:
                if attr_lower in _COOKIE_BOOL_ATTRS:
                    # Boolean attributes are set to True if present
                    morsel[attr_lower] = True
                else:
                    # Check if the attribute value contains another attribute
                    # This handles malformed cookies like "Path=/ Max-Age=123"
                    if attribute_value and " " in attribute_value:
                        # Look for known attributes in the value
                        for known_attr in _COOKIE_KNOWN_ATTRS:
                            pattern = f" {known_attr}="
                            pattern_idx = attribute_value.lower().find(pattern.lower())
                            if pattern_idx != -1:
                                # Split the value at this point
                                actual_value = attribute_value[:pattern_idx].strip()
                                morsel[attr_lower] = (
                                    _unquote(actual_value) if actual_value else ""
                                )

                                # Parse the remaining part as a new attribute
                                remaining = attribute_value[pattern_idx + 1 :]
                                remaining_eq_pos = remaining.find("=")
                                if remaining_eq_pos != -1:
                                    remaining_attr = (
                                        remaining[:remaining_eq_pos].strip().lower()
                                    )
                                    remaining_value = remaining[
                                        remaining_eq_pos + 1 :
                                    ].strip()
                                    if remaining_attr in _COOKIE_KNOWN_ATTRS:
                                        if remaining_attr in _COOKIE_BOOL_ATTRS:
                                            morsel[remaining_attr] = True
                                        else:
                                            morsel[remaining_attr] = (
                                                _unquote(remaining_value)
                                                if remaining_value
                                                else ""
                                            )
                                break
                        else:
                            # No known attribute found in value, use as-is
                            morsel[attr_lower] = (
                                _unquote(attribute_value) if attribute_value else ""
                            )
                    else:
                        # Set the attribute value, unquoting if necessary
                        morsel[attr_lower] = (
                            _unquote(attribute_value) if attribute_value else ""
                        )

    return (name, morsel)
```


Not that the regex was much better

Comment (bdraco) [reply aiohttp/helpers.py]: If I do the simple parser for Cookie, these tests fail:
tests/test_web_request.py::test_request_cookies_quoted_values FAILED                                                                                                                                                                         [ 35%]
tests/test_web_request.py::test_request_cookies_with_attributes FAILED                                                                                                                                                                       [ 35%]

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: > I don't see how we can ever make https://datatracker.ietf.org/doc/html/rfc6265#section-5.2 work for us as we specifically have tests for missing semi-colons and all sort of other issues

For context, there are 3 specs for cookies, the second one introduced the idea of quoted cookies. The problem is that nobody (e.g. no browser ever) implemented this behaviour and the spec was also incompatible with both the orginal spec and actual browser behaviour.

Python made the mistake of implementing that spec for http.cookies, thus making it incompatible with actual browsers. We've adopted that so far too, thus all the cookie-related bug reports we get.

The latest spec basically just tries to define cookies that are compatible with both of the previous specs and browsers, which means it's very broad and vague. You'll note that the syntax allows quotes around the cookie value in order to be compatible with the second spec, but doesn't actually define any concept of a quoted cookie.

So, basically, all the quoted cookie stuff should be abandoned by us. All the tests relating to it should be removed. Any other option will result in further bug reports of incompatibilities with browsers and real-world servers.

So, yeah, it's a bit of a rabbit hole.

Comment (bdraco) [reply aiohttp/helpers.py]: This isn't going to work. I'm going to revert back to the regex. Its ugly but less ugly than this abomination https://github.com/aio-libs/aiohttp/pull/11112#discussion_r2119624862

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: > I'm going to revert back to the regex

If you need to backport it, sure. But, if we want to fix this properly in v4 or something, then we need to drop all the quoted stuff.

Comment (bdraco) [reply aiohttp/helpers.py]: Yes, we need to backport this. It's addressing real-world breakage that needs to be fixed.

Comment (Dreamsorcerer) [reply aiohttp/helpers.py]: > Yes, we need to backport this. It's addressing real-world breakage that needs to be fixed.

They're all real-world breakage, but I'm pretty sure there is no possible way to fix all of them and not remove quoted cookies.

I think #5397 is the one I spent days or weeks on, but could never get it to work (tests in #5398).

Comment (bdraco) [reply aiohttp/helpers.py]: Thanks for the hint. Let me pull in the tests from https://github.com/aio-libs/aiohttp/pull/5398 and see what happens

Comment (bdraco) [reply aiohttp/helpers.py]: Part of the issue is shared cookies don't preserve the coded value but non-shared do. Our treatment there is wholly inconsistent

Comment (bdraco) [reply aiohttp/helpers.py]: I'll make a helper so treat them the same. That will fix the case in #5398

Comment (bdraco) [reply aiohttp/helpers.py]: Once I apply the same fix to the shared cookie path as https://github.com/aio-libs/aiohttp/pull/1453 the test works

File: aiohttp/helpers.py [hunk 2]
```diff
@@ -71,12 +73,77 @@
...
-__all__ = ("BasicAuth", "ChainMapProxy", "ETag", "frozen_dataclass_decorator", "reify")
+__all__ = (
+    "BasicAuth",
+    "ChainMapProxy",
+    "ETag",
+    "frozen_dataclass_decorator",
+    "make_non_quoted_morsel",
+    "make_quoted_morsel",
+    "preserve_morsel_with_coded_value",
+    "reify",
```
Comment (bdraco) [inline aiohttp/helpers.py]: not a good idea, someone will think this is ok for public use. Let me move them all to a protected namespace

File: tests/test_client_session.py [hunk 3]
```diff
@@ -758,7 +758,7 @@ async def handler(request: web.Request) -> web.Response:
     # getting the request URL as only parameter
     jar._filter_cookies_mock.assert_called_with(URL(req_url))
 
-    # Updating the cookie jar with the response cookies
+    # Updating the cookie jar with the cookies
```
Comment (bdraco) [inline tests/test_client_session.py]:
```suggestion
    # Updating the cookie jar with the response cookies
```

**Comment by bdraco** [comment]
Ready to go from my perspective. I want to sleep on it though

**Comment by bdraco** [comment]
Still happy with this after sleeping on it.  Just landing in LHR now to make the OHF summit in Dublin.  While I've done everything I can think of to write tests for it, I'm thinking it would be best to do a 3.12.7rc0 with this and run it through Home Assistant CI and a few other projects before pushing 3.12.7.

**Comment by bdraco** [comment]
I've done all the production testing I can do with the RC. All looks good

**Comment by Dreamsorcerer** [comment]
@bdraco We might be able to revert these changes in 3.15?
https://github.com/python/cpython/pull/113663

**Comment by bdraco** [comment]
Nice. I guess we can go back to the standard one in ~5.5 years 😄
